Source/deviceinfo/DeviceInfo.cpp

```diff
@@ -483,4 +483,151 @@ uint32_t deviceinfo_hdcp(deviceinfo_hdcp_t* supportedHDCP)
     }
     return result;
 }
+
+
+uint32_t deviceinfo_model_name(char buffer[], uint8_t* length)
+{
+    uint32_t result = Core::ERROR_UNAVAILABLE;
+
+    string modelName="";
+
+    Exchange::IDeviceCapabilities* iDeviceCapabilitiesPtr = DeviceInfoLink::Instance().Interface();
+    
+    if (iDeviceCapabilitiesPtr != nullptr) {
+        Exchange::IDeviceMetadata* iDeviceMetaDataPtr = iDeviceCapabilitiesPtr->QueryInterface<Exchange::IDeviceMetadata>();
```

Comment:
Now we do a query interface every time we need to get the interface which is rather expensive. It would be better to do this only one (and now we have the SmartInterfaceType class that one can keep it alive). You can check in the file DisplayInfo.cpp in the DisplayInfo client library how that is done, they solve a similar problem with a parent interface and a QueryInterface they need to for some other info while using the SmartInterfaceType to keep both alive.
(of course as an alternative we could get and cache most of the values once as they will never change on a running system. But that is something the original creator already could have done. Let's not do that right now, I'll create a jira issue to have that done later.)

Reply:
I created class data members holding pointers of the different interfaces. I did not use any parental class though because i think there is no need for it.